File: ccbusDemo/client/ccbusdemo/src/components/Downloader.js
import {Error} from 'ccbus';
import {ExtComponent} from 'ccbus';
import {WorkerServer} from 'ccbus';
import {FileSaver} from 'ccbus/lib/FileSaver';
import {FileDownload} from 'ccbus/payload/FileDownload';
import {FileUpload} from 'ccbus/payload/FileUpload';
import React from 'react';


/**
* @extends {ExtComponent}
*/ 
export class Downloader extends ExtComponent
{
    
    /**
    * @param {Event} e
    * @returns {void}
    */ 
    onClick(e)
    {
        this.download();
    }
    
    /**
    * @returns {void}
    */ 
    download()
    {
        WorkerServer.compute({
            onError: function(error)
            {
            },
            onCompletion: function(fileDownload)
            {
                FileSaver.saveAs(fileDownload.fileInputStream,fileDownload.fileName);
            },
            
            /**
            * @type {String}
            */ 
            endpoint: "Downloader_download_0",
            
            /**
            * @type {String}
            */ 
            responseType: "blob",
            
            /**
            * @type {Object}
            */ 
            data: {
                
            },
            
            /**
            * @type {Object}
            */ 
            bind: this
        });
    }

    render()
    {
        return(<span><button onClick={(e)=>{this.onClick(e)}}>download</button></span>)
    }
}

File: ccbusDemo/server/ccbusdemo/src/main/java/ccbus/demo/ccbus/web/components/Downloader.java
package ccbus.demo.ccbus.web.components;

import ccbus.connect.core.ccbus.Error;
import ccbus.connect.core.ccbus.ExtComponent;
import ccbus.connect.core.ccbus.WorkerServer;
import ccbus.connect.core.ccbus.lib.FileSaver;
import ccbus.connect.core.ccbus.payload.FileDownload;
import ccbus.connect.ecma.Event;

import java.io.File;

public class Downloader extends ExtComponent
{

    void onClick(Event e)
    {
        download();
    }

    void download()
    {
        new WorkerServer<FileDownload>()
        {
            @Override
            public void compute() {
                String fileName="Screenshot_20200108.png";
                File file = new File(this.getClass().getClassLoader().getResource("public/pics/").getPath()+fileName);
                complete(new FileDownload(file));
            }

            @Override
            public void onCompletion(FileDownload fileDownload)
            {
                FileSaver.saveAs(fileDownload.fileInputStream,fileDownload.fileName);
            }

            @Override
            public void onError(Error error) {

            }
        };
    }
}
